src/lightning/data/cache/dataloader.py

```diff
@@ -0,0 +1,163 @@
+
+    def __len__(self) -> int:
+        return len(self.cache) if self.cache.filled else len(self._datataset)
+
+    def __getitem__(self, index):
+        data = self.cache[index] if self.cache.filled else self._datataset[index]
+        if not self.cache.filled:
+            self.cache[index] = data
+        return data
+
+
+class CacheCollateFn:
+    def __init__(self):
+        self.collate_fn = default_collate
+
+    def __call__(self, items):
+        if all(item is None for item in items):
+            return None
+        return self.collate_fn(items)
+
+
+class _SingleProcessDataLoaderIterPatch(_SingleProcessDataLoaderIter):
+    """This is overriden to inform the cache is done chunking."""
+
+    def _next_data(self):
+        try:
+            return super()._next_data()
+        except StopIteration:
+            for v in self._dataset_fetcher.dataset.__dict__.values():
+                if isinstance(v, Cache):
+                    v.done()
+            raise StopIteration()
+
+
+class _MultiProcessingDataLoaderIterPatch(_MultiProcessingDataLoaderIter):
+    def __init__(self, loader):
+        # Patch PyTorch worker loop
+        from torch.utils.data._utils import worker
+
+        from lightning.data.cache.worker import _worker_loop
+
+        worker._worker_loop = _worker_loop
+        super().__init__(loader)
+
+
+class CacheDataLoader(DataLoader):
+    def __init__(
+        self,
+        dataset,
+        *args,
+        sampler=None,
+        batch_sampler=None,
+        num_workers=0,
+        shuffle: bool = False,
+        generator=None,
+        batch_size=1,
+        drop_last=False,
+        cache_dir: Optional[str] = None,
+        chunk_size: Optional[int] = 2 << 26,
+        compression: Optional[str] = None,
+        **kwargs,
+    ):
+        if sampler:
+            raise Exception("Passing a sampler isn't supported with the CacheDataLoader.")
+
+        if batch_sampler:
+            raise Exception("Passing a batch_sampler isn't supported with the CacheDataLoader.")
+
+        if isinstance(dataset, IterableDataset):
+            raise Exception("Only map-based dataset are supported by the CacheDataLoader for now.")
+
+        cache = [v for v in dataset.__dict__.values() if isinstance(v, Cache)]
+
+        if len(cache) > 1:
+            raise Exception(f"The CacheDataloader should be used with a dataset using a single cache. Found {cache}.")
+
+        if len(cache) == 0:
+            if cache_dir is None:
+                logger.info("You can provide a `cache_dir` filepath to the CacheDataLoader.")
+
+            dataset = CacheDataset(dataset, cache_dir, chunk_size, compression)
+            cache = dataset.cache
+        else:
+            cache = cache[0]
+
+        cache._setup(num_workers)
+
+        if not cache.filled and shuffle:
+            logger.info("Shuffle is ignored during caching phase")
+
+        distributed_env = _DistributedEnv.detect()
+        batch_sampler = CacheBatchSampler(
```

Comment:
There is 2 approaches:
1) Auto wrap where we cache whatever the user dataset returns including data augmentation.
2) The user utilises the cache himself and decide where to apply the cache (this would be the recommended one)

Reply:
@lantiga Done.